File: packages/go-sdk/turbosign_sender_config_test.go
package turbodocx

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Helper function to parse form data from request
func parseFormData(t *testing.T, r *http.Request) map[string]string {
	contentType := r.Header.Get("Content-Type")
	formData := make(map[string]string)

	if strings.Contains(contentType, "multipart/form-data") {
		err := r.ParseMultipartForm(10 << 20) // 10MB max
		require.NoError(t, err)
		for key, values := range r.MultipartForm.Value {
			if len(values) > 0 {
				formData[key] = values[0]
			}
		}
	} else if strings.Contains(contentType, "application/json") {
		body, err := io.ReadAll(r.Body)
		require.NoError(t, err)
		var jsonData map[string]interface{}
		err = json.Unmarshal(body, &jsonData)
		require.NoError(t, err)
		for key, value := range jsonData {
			if str, ok := value.(string); ok {
				formData[key] = str
			}
		}
	}

	return formData
}

// Tests for CreateSignatureReviewLink with configured sender
func TestTurboSign_CreateSignatureReviewLink_ConfiguredSender(t *testing.T) {
	t.Run("should use configured senderEmail when not provided in request", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"status":     "review_ready",
				"message":    "Document prepared",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
			// senderEmail and senderName NOT provided
		})

		require.NoError(t, err)
		assert.Equal(t, "configured@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Configured Support", capturedFormData["senderName"])
	})

	t.Run("should use configured senderEmail only when senderName not configured", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"status":     "review_ready",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			// senderName NOT configured
			BaseURL: server.URL,
		})

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
		})

		require.NoError(t, err)
		assert.Equal(t, "configured@company.com", capturedFormData["senderEmail"])
		_, hasSenderName := capturedFormData["senderName"]
		assert.False(t, hasSenderName, "senderName should not be in form data when not configured")
	})

	t.Run("should override configured sender with request-level sender", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"status":     "review_ready",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
			SenderEmail: "override@company.com",
			SenderName:  "Override Support",
		})

		require.NoError(t, err)
		assert.Equal(t, "override@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Override Support", capturedFormData["senderName"])
	})

	t.Run("should partially override - use request senderEmail but configured senderName", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"status":     "review_ready",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
			SenderEmail: "override@company.com",
			// senderName NOT provided - should use configured value
		})

		require.NoError(t, err)
		assert.Equal(t, "override@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Configured Support", capturedFormData["senderName"])
	})
}

// Tests for SendSignature with configured sender
func TestTurboSign_SendSignature_ConfiguredSender(t *testing.T) {
	t.Run("should use configured senderEmail and senderName", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"message":    "Document sent for signing",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		_, err := client.TurboSign.SendSignature(context.Background(), &SendSignatureRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
		})

		require.NoError(t, err)
		assert.Equal(t, "configured@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Configured Support", capturedFormData["senderName"])
	})

	t.Run("should allow request-level override in sendSignature", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-123",
				"message":    "Document sent for signing",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		_, err := client.TurboSign.SendSignature(context.Background(), &SendSignatureRequest{
			FileLink: "https://example.com/doc.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
			SenderEmail: "sales@company.com",
			SenderName:  "Sales Team",
		})

		require.NoError(t, err)
		assert.Equal(t, "sales@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Sales Team", capturedFormData["senderName"])
	})
}

// Tests for file upload with configured sender
func TestTurboSign_FileUpload_ConfiguredSender(t *testing.T) {
	t.Run("should use configured sender in file upload requests", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-upload",
				"status":     "review_ready",
				"message":    "Document prepared",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		mockFile := []byte("%PDF-mock-content")

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			File:     mockFile,
			FileName: "document.pdf",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
		})

		require.NoError(t, err)
		assert.Equal(t, "configured@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Configured Support", capturedFormData["senderName"])
	})

	t.Run("should override configured sender in file upload requests", func(t *testing.T) {
		var capturedFormData map[string]string

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			capturedFormData = parseFormData(t, r)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(map[string]interface{}{
				"success":    true,
				"documentId": "doc-upload",
				"status":     "review_ready",
				"message":    "Document prepared",
			})
		}))
		defer server.Close()

		client, _ := NewClientWithConfig(ClientConfig{
			APIKey:      "test-api-key",
			OrgID:       "test-org-id",
			SenderEmail: "configured@company.com",
			SenderName:  "Configured Support",
			BaseURL:     server.URL,
		})

		mockFile := []byte("%PDF-mock-content")

		_, err := client.TurboSign.CreateSignatureReviewLink(context.Background(), &CreateSignatureReviewLinkRequest{
			File:        mockFile,
			FileName:    "document.pdf",
			SenderEmail: "specific@company.com",
			SenderName:  "Specific Team",
			Recipients: []Recipient{
				{Name: "John Doe", Email: "john@example.com", SigningOrder: 1},
			},
			Fields: []Field{
				{Type: "signature", Page: 1, X: 100, Y: 500, Width: 200, Height: 50, RecipientEmail: "john@example.com"},
			},
		})

		require.NoError(t, err)
		assert.Equal(t, "specific@company.com", capturedFormData["senderEmail"])
		assert.Equal(t, "Specific Team", capturedFormData["senderName"])
	})
}

File: packages/java-sdk/src/test/java/com/turbodocx/TurboSignSenderConfigTest.java
package com.turbodocx;

import com.google.gson.Gson;
import com.google.gson.JsonParser;
import com.turbodocx.models.*;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * TurboSign Sender Configuration Tests
 *
 * Tests to ensure sender_email/sender_name from configuration are properly used
 * in signature requests, with per-request override capability
 */
class TurboSignSenderConfigTest {

    private MockWebServer server;
    private TurboDocxClient client;
    private final Gson gson = new Gson();

    @BeforeEach
    void setUp() throws IOException {
        server = new MockWebServer();
        server.start();

        client = new TurboDocxClient.Builder()
                .apiKey("test-api-key")
                .orgId("test-org-id")
                .senderEmail("configured@company.com")
                .senderName("Configured Support")
                .baseUrl(server.url("/").toString())
                .build();
    }

    @AfterEach
    void tearDown() throws IOException {
        server.shutdown();
    }

    /**
     * Helper method to extract form data from multipart or JSON request
     */
    private Map<String, String> extractFormData(RecordedRequest request) throws Exception {
        Map<String, String> formData = new HashMap<>();
        String contentType = request.getHeader("Content-Type");

        if (contentType != null && contentType.contains("multipart/form-data")) {
            // Parse multipart form data
            String body = request.getBody().readString(StandardCharsets.UTF_8);
            String[] parts = body.split("--");

            for (String part : parts) {
                if (part.contains("Content-Disposition: form-data; name=\"")) {
                    int nameStart = part.indexOf("name=\"") + 6;
                    int nameEnd = part.indexOf("\"", nameStart);
                    if (nameEnd > nameStart) {
                        String name = part.substring(nameStart, nameEnd);

                        // Find the value after the headers
                        int valueStart = part.indexOf("\r\n\r\n");
                        if (valueStart > 0) {
                            valueStart += 4; // Skip \r\n\r\n
                            int valueEnd = part.lastIndexOf("\r\n");
                            if (valueEnd > valueStart) {
                                String value = part.substring(valueStart, valueEnd);
                                formData.put(name, value);
                            }
                        }
                    }
                }
            }
        } else if (contentType != null && contentType.contains("application/json")) {
            // Parse JSON body
            String body = request.getBody().readString(StandardCharsets.UTF_8);
            var jsonObject = JsonParser.parseString(body).getAsJsonObject();
            for (String key : jsonObject.keySet()) {
                formData.put(key, jsonObject.get(key).getAsString());
            }
        }

        return formData;
    }

    // ============================================
    // CreateSignatureReviewLink with Configured Sender
    // ============================================

    @Test
    @DisplayName("should use configured senderEmail when not provided in request")
    void shouldUseConfiguredSenderEmailWhenNotProvided() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "status", "review_ready",
                        "message", "Document prepared"
                ))));

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                // senderEmail and senderName NOT provided
                .build();

        client.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("configured@company.com", formData.get("senderEmail"));
        assertEquals("Configured Support", formData.get("senderName"));
    }

    @Test
    @DisplayName("should use configured senderEmail only when senderName not configured")
    void shouldUseConfiguredSenderEmailOnlyWhenSenderNameNotConfigured() throws Exception {
        server.shutdown();
        server = new MockWebServer();
        server.start();

        // Create client WITHOUT senderName
        TurboDocxClient clientWithoutName = new TurboDocxClient.Builder()
                .apiKey("test-api-key")
                .orgId("test-org-id")
                .senderEmail("configured@company.com")
                // senderName NOT configured
                .baseUrl(server.url("/").toString())
                .build();

        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "status", "review_ready"
                ))));

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .build();

        clientWithoutName.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("configured@company.com", formData.get("senderEmail"));
        assertNull(formData.get("senderName"), "senderName should not be in form data when not configured");
    }

    @Test
    @DisplayName("should override configured sender with request-level sender")
    void shouldOverrideConfiguredSenderWithRequestLevelSender() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "status", "review_ready"
                ))));

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .senderEmail("override@company.com")
                .senderName("Override Support")
                .build();

        client.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("override@company.com", formData.get("senderEmail"));
        assertEquals("Override Support", formData.get("senderName"));
    }

    @Test
    @DisplayName("should partially override - use request senderEmail but configured senderName")
    void shouldPartiallyOverrideUseRequestSenderEmailButConfiguredSenderName() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "status", "review_ready"
                ))));

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .senderEmail("override@company.com")
                // senderName NOT provided - should use configured value
                .build();

        client.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("override@company.com", formData.get("senderEmail"));
        assertEquals("Configured Support", formData.get("senderName"));
    }

    // ============================================
    // SendSignature with Configured Sender
    // ============================================

    @Test
    @DisplayName("should use configured senderEmail and senderName in sendSignature")
    void shouldUseConfiguredSenderEmailAndSenderNameInSendSignature() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "message", "Document sent for signing",
                        "recipients", Collections.emptyList()
                ))));

        SendSignatureRequest request = new SendSignatureRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .build();

        client.turboSign().sendSignature(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("configured@company.com", formData.get("senderEmail"));
        assertEquals("Configured Support", formData.get("senderName"));
    }

    @Test
    @DisplayName("should allow request-level override in sendSignature")
    void shouldAllowRequestLevelOverrideInSendSignature() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-123",
                        "message", "Document sent for signing",
                        "recipients", Collections.emptyList()
                ))));

        SendSignatureRequest request = new SendSignatureRequest.Builder()
                .fileLink("https://example.com/doc.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .senderEmail("sales@company.com")
                .senderName("Sales Team")
                .build();

        client.turboSign().sendSignature(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("sales@company.com", formData.get("senderEmail"));
        assertEquals("Sales Team", formData.get("senderName"));
    }

    // ============================================
    // File Upload with Configured Sender
    // ============================================

    @Test
    @DisplayName("should use configured sender in file upload requests")
    void shouldUseConfiguredSenderInFileUploadRequests() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-upload",
                        "status", "review_ready",
                        "message", "Document prepared"
                ))));

        byte[] mockFile = new byte[]{0x25, 0x50, 0x44, 0x46}; // %PDF

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .file(mockFile)
                .fileName("document.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .build();

        client.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("configured@company.com", formData.get("senderEmail"));
        assertEquals("Configured Support", formData.get("senderName"));
    }

    @Test
    @DisplayName("should override configured sender in file upload requests")
    void shouldOverrideConfiguredSenderInFileUploadRequests() throws Exception {
        server.enqueue(new MockResponse()
                .setResponseCode(200)
                .setHeader("Content-Type", "application/json")
                .setBody(gson.toJson(Map.of(
                        "success", true,
                        "documentId", "doc-upload",
                        "status", "review_ready",
                        "message", "Document prepared"
                ))));

        byte[] mockFile = new byte[]{0x25, 0x50, 0x44, 0x46}; // %PDF

        CreateSignatureReviewLinkRequest request = new CreateSignatureReviewLinkRequest.Builder()
                .file(mockFile)
                .fileName("document.pdf")
                .recipients(Collections.singletonList(
                        new Recipient("John Doe", "john@example.com", 1)))
                .fields(Collections.singletonList(
                        new Field("signature", 1, 100, 500, 200, 50, "john@example.com")))
                .senderEmail("specific@company.com")
                .senderName("Specific Team")
                .build();

        client.turboSign().createSignatureReviewLink(request);

        RecordedRequest recorded = server.takeRequest();
        Map<String, String> formData = extractFormData(recorded);

        assertEquals("specific@company.com", formData.get("senderEmail"));
        assertEquals("Specific Team", formData.get("senderName"));
    }
}
